Framework/API/inc/MantidAPI/MatrixWorkspace.h

```diff
@@ -551,9 +554,14 @@ class MANTID_API_DLL MatrixWorkspace : public IMDWorkspace,
                     const std::size_t &YLength) = 0;
   virtual void init(const HistogramData::Histogram &histogram) = 0;
 
+  virtual ISpectrum &getSpectrumWithoutInvalidation(const size_t index) = 0;
+
   /// Invalidates the commons bins flag.  This is generally called when a method
   /// could allow the X values to be changed.
-  void invalidateCommonBinsFlag() { m_isCommonBinsFlagSet = false; }
+  void invalidateCommonBinsFlag() {
+    std::lock_guard<std::mutex> lock(m_commonBinsMutex);
```

Comment:
Isn't this basically putting a lock into the loop over spectra in *all* algorithms modifying X? This sounds quite expensive too me, did you check if this is visible in any of the performance tests?

Edit: Actually it is even worse. `invalidateCommonBinsFlag` is effectively also called from any method modifying Y or E (this could potentially be avoided), so we get this lock in all of our algorithms. :(

Reply:
I'll run some performance tests to get an idea how much impact the lock has.

Reply:
By the way, `mutableY`, `mutableE`, `mutableDx`, `dataY`, `dataE` and `dataDx` call `getSpectrumWithoutInvalidation` which skips the locking.

Reply:
True that alleviates the problem slightly. Nevertheless, our code frequently uses something like

```cpp
auto &spec = ws.getSpectrum(i);
auto &y = spec.y();
```

and I think that will lead to invalidation?

Reply:
Yes. All such calls to non-const `getSpectrum` which potentially lead to modifiable X invalidate the flag.

Reply:
There is a definitive performance impact in code like (from `Workspace2DTestPerformance`)
```c++
for (size_t i = 0; i < ws1->getNumberHistograms(); i++) {
  auto &spec = ws1->getSpectrum(i);
  spec.setDetectorID(detid_t(i));
}
```
We could make `getSpectrumWithoutInvalidaton` public and non-const and use that in these cases. Perhaps with a better name as well. It might lead to some misuse, though. I'm all ears if you have other suggestions.

Reply:
Ultimately I guess this boils down to whether the performance gains from caching the flag are greater what we loose in locking. We might as well remove that thing and live with the fact that `isCommonBins()` and `CommonBinsValidator` are expensive operations.

Reply:
What about the attempt to use atomics in Martyn's PR that was linked above?

Reply:
Slightly off topic: https://github.com/mantidproject/dataset solves this "for free" by supporting true multi-dimensional data. The `isCommonBins` check is than as simple as `dataset[Coord.Tof].dimensions().contains(Dim::Spectrum)`.

Reply:
Using atomics for sure solves the performance problem. Why I ended up using a lock was because this prevented a thread from invalidating the flag while another thread was executing `isCommonBins()`. However, if we are executing `isCommonBins()` while changing the X values I guess we are already in the deep end of the pool, waking up the sleeping Great Old Ones.